src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs

```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
         {
-            var compilationResult = _compiler.Compile(actionDescriptor.RelativePath);
-            var compiledTypeInfo = compilationResult.CompiledType.GetTypeInfo();
-            // If a model type wasn't set in code then the model property's type will be the same
-            // as the compiled type.
-            var modelTypeInfo = compiledTypeInfo.GetProperty(ModelPropertyName)?.PropertyType.GetTypeInfo();
-            if (modelTypeInfo == compiledTypeInfo)
+            var result = _compiler.Compile(actionDescriptor.RelativePath);
+            return CreateDescriptor(actionDescriptor, result.CompiledType.GetTypeInfo());
+        }
+
+        // Internal for unit testing
+        internal static CompiledPageActionDescriptor CreateDescriptor(PageActionDescriptor actionDescriptor, TypeInfo pageType)
+        {
+            // Pages always have a model type. If it's not set explicitly by the developer using
+            // @model, it will be the same as the page type.
+            //
+            // However, we allow it to be null here for ease of testing.
+            var modelType = pageType.GetProperty(ModelPropertyName)?.PropertyType.GetTypeInfo();
+
+            // Now we want to find the handler methods. If the model defines any handlers, then we'll use those,
+            // otherwise look at the page itself (unless the page IS the model, in which case we already looked).
+            TypeInfo handlerType;
+
+            var handlerMethods = modelType == null ? null : CreateHandlerMethods(modelType);
+            if (handlerMethods != null && handlerMethods.Length > 0)
+            {
+                handlerType = modelType;
+            }
+            else
             {
-                modelTypeInfo = null;
+                handlerType = pageType;
+                handlerMethods = CreateHandlerMethods(pageType);
             }
 
+            var boundProperties = CreateBoundProperties(handlerType);
+            
             return new CompiledPageActionDescriptor(actionDescriptor)
             {
-                PageTypeInfo = compiledTypeInfo,
-                ModelTypeInfo = modelTypeInfo,
+                BoundProperties = boundProperties,
+                HandlerMethods = handlerMethods,
+                HandlerTypeInfo = handlerType,
+                ModelTypeInfo = modelType,
+                PageTypeInfo = pageType,
             };
         }
+
+        internal static HandlerMethodDescriptor[] CreateHandlerMethods(TypeInfo type)
+        {
+            var methods = type.GetMethods();
+            var results = new HandlerMethodDescriptor[methods.Length];
+
+            for (var i = 0; i < methods.Length; i++)
+            {
+                var method = methods[i];
+                if (!IsValidHandlerMethod(method))
+                {
+                    continue;
+                }
+
+                if (method.IsDefined(typeof(NonHandlerAttribute)))
```

Comment:
Move these two checks inside the `IsValidHandlerMethod` (to be consistent with what we do in https://github.com/aspnet/Mvc/blob/0cee00aae1a4064aa2ab2608015dac92d6389c2e/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs#L381